lib/translator/mode_sense.cc

```diff
@@ -0,0 +1,389 @@
+// limitations under the License.
+
+#include "mode_sense.h"
+
+namespace translator {
+
+namespace {
+
+// Section 6.3.3.1
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+constexpr scsi::CachingModePage caching_mode_page = {
+    .ps = 0,
+    .spf = 0,
+    .page_code = scsi::ModePageCode::kCacheMode,
+    .page_length = 0x12,
+    .ic = 1,
+};
+
+// Section 6.3.3.2
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+constexpr scsi::ControlModePage control_mode_page = {
+    .ps = 0,
+    .spf = 0,
+    .page_code = scsi::ModePageCode::kControlMode,
+    .page_length = 0x0a,
+    .tst = 0,
+    .tmf_only = 0,
+    .dpicz = 0,
+    .d_sense = 1,
+    .gltsd = 1,
+    .rlec = 0,
+    .qam = 0,
+    .nuar = 0,
+    .qerr = 0x01,
+    .rac = 0,
+    .ua_intlck_ctrl = 0,
+    .swp = 0,
+    .ato = 0,
+    .tas = 0x01,
+    .autoload_mode = 0,
+    .busy_timeout_period = 0xFFFF,
+    .estct = 0};
+
+// Section 6.3.3.3
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+constexpr scsi::PowerConditionModePage power_condition_mode_page = {
+    .ps = 0,
+    .spf = 0,
+    .page_code = scsi::ModePageCode::kPowerConditionMode,
+    .page_length = 0x26,
+    .pm_bg_precedence = 0};
+
+struct CommonCmdAttributes {
+  scsi::ModePageCode page_code;
+  scsi::PageControl pc;
+  bool dbd;    // Disable block descriptors
+  bool llbaa;  // Long LBA accepted
+};
+
+// Generates an nvme identify command for fetching block descriptor data
+StatusCode GenerateBlockDescriptorIdentifyCmd(
+    nvme::GenericQueueEntryCmd& nvme_cmd, Allocation& allocation,
+    uint32_t nsid) {
+  nvme_cmd = {.opc = static_cast<uint8_t>(nvme::AdminOpcode::kIdentify),
+              .nsid = nsid};
+
+  if (allocation.SetPages(1, 0) == StatusCode::kFailure)
+    return StatusCode::kFailure;
+  nvme_cmd.dptr.prp.prp1 = allocation.data_addr;
+}
+
+// Generates an nvme get features command for fetching cache features
+StatusCode GenerateCacheGetFeaturesCmd(scsi::PageControl pc, uint32_t nsid,
+                                       nvme::GenericQueueEntryCmd& nvme_cmd) {
+  nvme::GetFeaturesCmd tmp_get_features = {};
+  tmp_get_features.opc = static_cast<uint8_t>(nvme::AdminOpcode::kGetFeatures);
+  tmp_get_features.nsid = nsid;
+  switch (pc) {
+    case scsi::PageControl::kCurrent:
+      tmp_get_features.sel = nvme::FeatureSelect::kCurrent;
+      break;
+    case scsi::PageControl::kChangeable:
+      tmp_get_features.sel = nvme::FeatureSelect::kSaved;
+      break;
+    case scsi::PageControl::kDefault:
+      tmp_get_features.sel = nvme::FeatureSelect::kDefault;
+      break;
+    default:
+      DebugLog("Unsupported page control recieved");
+      return StatusCode::kFailure;
+  }
+  tmp_get_features.fid = nvme::FeatureType::kVolatileWriteCache;
+  memcpy(&nvme_cmd, &tmp_get_features, sizeof(nvme::GenericQueueEntryCmd));
```

Comment:
readval

Reply:
Can't readval here because neither types are a span. We can provide more type safety with static_assert(sizeof(nvme_cmd) == sizeof(tmp_get_features))

Reply:
You could cast the pointer to a byte array and construct a span over that, using sizeof() to determine the length.

Reply:
Does doing this actually provide any further safety or is it aesthetic?

Reply:
I agree that the benefit is dubious. Given the static assert I'd leave it as-is for simplicity.